components/application-mgt/org.wso2.carbon.identity.application.mgt/src/main/java/org/wso2/carbon/identity/application/mgt/provider/RegistryBasedApplicationPermissionProvider.java

```diff
@@ -0,0 +1,314 @@
+import org.wso2.carbon.identity.application.common.model.PermissionsAndRoleConfig;
+import org.wso2.carbon.identity.application.mgt.ApplicationMgtUtil;
+import org.wso2.carbon.identity.base.IdentityException;
+import org.wso2.carbon.identity.core.util.IdentityTenantUtil;
+import org.wso2.carbon.registry.api.Collection;
+import org.wso2.carbon.registry.api.Registry;
+import org.wso2.carbon.registry.api.RegistryException;
+import org.wso2.carbon.registry.api.Resource;
+import org.wso2.carbon.registry.core.RegistryConstants;
+import org.wso2.carbon.user.core.UserRealm;
+import org.wso2.carbon.user.mgt.UserMgtConstants;
+import org.wso2.carbon.utils.multitenancy.MultitenantConstants;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.Objects;
+
+import static org.wso2.carbon.identity.application.mgt.ApplicationMgtUtil.PATH_CONSTANT;
+
+/**
+ * Registry based application permission provider class.
+ */
+public class RegistryBasedApplicationPermissionProvider implements ApplicationPermissionProvider {
+
+    private static final int MAX_RETRY_ATTEMPTS = 3;
+    private static final String APPLICATION_ROOT_PERMISSION = "applications";
+
+    private static Log log = LogFactory.getLog(RegistryBasedApplicationPermissionProvider.class);
+
+    @Override
+    public void renameAppPermissionPathNode(String oldName, String newName)
+            throws IdentityApplicationManagementException {
+
+        List<ApplicationPermission> loadPermissions = loadPermissions(oldName);
+        String newApplicationNode = ApplicationMgtUtil.getApplicationPermissionPath() + PATH_CONSTANT + oldName;
+        Registry tenantGovReg = CarbonContext.getThreadLocalCarbonContext().getRegistry(RegistryType.USER_GOVERNANCE);
+        // Creating new application node.
+        try {
+            for (ApplicationPermission applicationPermission : loadPermissions) {
+                tenantGovReg.delete(newApplicationNode + PATH_CONSTANT + applicationPermission.getValue());
+            }
+            tenantGovReg.delete(newApplicationNode);
+            Collection permissionNode = tenantGovReg.newCollection();
+            permissionNode.setProperty("name", newName);
+            newApplicationNode = ApplicationMgtUtil.getApplicationPermissionPath() + PATH_CONSTANT + newName;
+            String applicationNode = newApplicationNode;
+            tenantGovReg.put(newApplicationNode, permissionNode);
+            addPermission(applicationNode, loadPermissions.toArray(new ApplicationPermission[loadPermissions.size()]),
+                    tenantGovReg);
+        } catch (RegistryException e) {
+            throw new IdentityApplicationManagementException(
+                    "Error while renaming permission node " + oldName + "to " + newName, e);
+        }
+    }
+
+    @Override
+    public void storePermissions(String applicationName, String username, PermissionsAndRoleConfig permissionsConfig)
```

Comment:
Can we remove the username from the method argument? because it won't make any sense for the non-registry provider.   

We are using this argument to check whether the user is authorized to create the resource path in the registry. Other than that we can always use the `realm.getRealmConfiguration().getAdminUserName()` and then finally reset the user. this way we don't need to pass the username.

Reply:
Fixed: a55181d5fdd11acf6f23808651ab50253ad85368

I read the username from PrivilegedCarbonContext instead of passing as an argument.